### PR TITLE
Remove restriction

### PR DESCRIPTION
**Pull request by young917**
### **Remove restriction of length of seq(=7)**

* **Before pass input to the Encoder or Decoder** (def controller_train / controller_infer / generate_synthetic_controller_data):

sort seqs by length of seq in one batch. 

* **Decoder**: 

pack_padded_sequence before rnn : make rnn ignore the inputs outside the length of the seq.

pad_packed_seqeunce after rnn : restore original shape.

pass mask to attention model : Because the length of seqs can be different each other in one batch, use mask to ignore out-of-length.

* **Encoder**:

pack_padded_sequence before rnn :  make rnn ignore the inputs outside the length of the seq.

pad_packed_seqeunce after rnn : restore original shape

* **DataLoader**: 

add 'input_len' (list of length of seqs in one batch) -> pass to the encoder / decoder

* compatible with our nasbench api

### CODE REVIEW

**Comment by young917** [comment]
Checked that the loss is reducing.
Checked that the final results include seq with length < 7.

File: nasbench/decoder.py [hunk 1]
```diff
@@ -73,25 +75,40 @@ def __init__(self,
...
+
             x = (residual + x) * math.sqrt(0.5)
             residual = x
-            x, _ = self.attention(x, encoder_outputs)
+            
+            # create mask
+            mask = torch.zeros(bsz, x.size(1))
+            for i,l in enumerate(x_len):
+                for j in range(l):
+                    mask[i][j] = 1
```
Comment (juice500ml) [inline nasbench/decoder.py:100]: What about
```
for i, l in enumerate(x_len):
  mask[i, :j] = 1
```
?

File: nasbench/utils.py [hunk 2]
```diff
@@ -48,19 +49,19 @@ def generate_arch(n, nasbench, need_perf=False):
     np.random.shuffle(all_keys)
     for key in all_keys:
         fixed_stat, computed_stat = nasbench.get_metrics_from_hash(key)
-        if len(fixed_stat['module_operations']) < 7:
-            continue
+        #if len(fixed_stat['module_operations']) < 7:
+        #    continue
```
Comment (juice500ml) [inline nasbench/utils.py:53]: Remove plz

File: nasbench/decoder.py [hunk 3]
```diff
@@ -73,25 +75,40 @@ def __init__(self,
         for i in range(self.n):
             self.offsets.append( (i + 3) * i // 2 - 1)
     
-    def forward(self, x, encoder_hidden=None, encoder_outputs=None):
+    def forward(self, x, x_len, encoder_hidden=None, encoder_outputs=None):
+        # x is decoder_inputs = [0] + encoder_inputs[:-1]
```
Comment (juice500ml) [inline nasbench/decoder.py:79]: Remove this

File: nasbench/decoder.py [hunk 4]
```diff
@@ -73,25 +75,40 @@ def __init__(self,
         for i in range(self.n):
             self.offsets.append( (i + 3) * i // 2 - 1)
     
-    def forward(self, x, encoder_hidden=None, encoder_outputs=None):
+    def forward(self, x, x_len, encoder_hidden=None, encoder_outputs=None):
```
Comment (juice500ml) [inline nasbench/decoder.py:78]: x_len feels like a somewhat misnomer. If i understood it correctly, it is somewhat like this: 
x_len = [len(x) for x in xs]
So, maybe x_len_per_elem? x_len_list? Or at least some comments would be helpful!

File: nasbench/decoder.py [hunk 5]
```diff
@@ -73,25 +75,40 @@ def __init__(self,
...
+                for j in range(l):
+                    mask[i][j] = 1
+            mask = (mask == 0).unsqueeze(1)
+            mask = utils.move_to_cuda(mask)
+            
+            x, _ = self.attention(x, encoder_outputs, mask=mask)
             x = (residual + x) * math.sqrt(0.5)
             predicted_softmax = F.log_softmax(self.out(x.view(-1, self.hidden_size)), dim=-1)
             predicted_softmax = predicted_softmax.view(bsz, tgt_len, -1)
             return predicted_softmax, None
```
Comment (juice500ml) [inline nasbench/decoder.py:108]: Is predicted softmax returns sane values? If padded elements are zero-initialized, then probabilty values will be broken.

**Review by juice500ml — CHANGES_REQUESTED** [review]
Updating this code to self-supervised-nas will take more time. Let's make sure this works!
